### PR TITLE
Updated BABE threshold to be configurable

### LINKED ISSUE #1233: update babe-threshold in config.toml to be babe-threshold-numerator and babe-threshold-denominator
## Expected Behavior

- for integration tests, it's handy to be able to modify the BABE threshold to allow for a certain rate of block production
- however currently there is only an option for setting the threshold either to the max or min value
- it would be handier to have this as a probability of a producer being allowed to produce a block, ie. C1 and C2 (numerator and denomator) in the babe configuration
- this way if we have a 3 node test, we can make it so around 1 block is produced for slot by setting C1 = 1 and C2 = 3 for example

## Checklist

- [x] I have read [CODE_OF_CONDUCT](https://github.com/ChainSafe/gossamer/blob/development/.github/CODE_OF_CONDUCT.md) and [CONTRIBUTING](https://github.com/ChainSafe/gossamer/blob/development/.github/CONTRIBUTING.md) 
- [x] I have provided as much information as possible and necessary
- [ ] I am planning to submit a pull request to fix this issue myself

### PR DESCRIPTION
**Pull request by RyRy79261**
## Changes

- Changed BabeThreshold to have a numerator & denominator config
- Changed setting logic for setting the threshold

## Tests

```
go test ./... -short
```

## Checklist

- [x] I have read [CODE_OF_CONDUCT](https://github.com/ChainSafe/gossamer/blob/development/.github/CODE_OF_CONDUCT.md) and [CONTRIBUTING](https://github.com/ChainSafe/gossamer/blob/development/.github/CONTRIBUTING.md) 
- [x] I have provided as much information as possible and necessary
- [ ] I have reviewed my own pull request before requesting a review
- [ ] All integration tests and required coverage checks are passing

## Issues

- closes #1233

### CODE REVIEW

File: dot/config.go [hunk 1]
```diff
@@ -83,13 +83,14 @@ type NetworkConfig struct {
...
-	GrandpaAuthority bool
-	BabeThreshold    *big.Int
-	SlotDuration     uint64
-	EpochLength      uint64
-	WasmInterpreter  string
+	Roles                    byte
+	BabeAuthority            bool
+	GrandpaAuthority         bool
+	BabeThresholdNumerator   *big.Int
+	BabeThresholdDenominator *big.Int
```
Comment (noot) [inline dot/config.go:90]: are `*big.Int`s able to be marshaled to/from toml?

File: dot/core/service.go [hunk 2]
```diff
@@ -96,8 +96,9 @@ type Config struct {
 	IsFinalityAuthority bool
 	Verifier            Verifier
 
-	NewBlocks     chan types.Block // only used for testing purposes
-	BabeThreshold *big.Int         // used by Verifier, for development purposes
+	NewBlocks                chan types.Block // only used for testing purposes
+	BabeThresholdNumerator   *big.Int         // used by Verifier, for development purposes
+	BabeThresholdDenominator *big.Int         // used by Verifier, for development purposes
```
Comment (noot) [inline dot/core/service.go:101]: core doesn't use this anymore, pls remove

File: dot/node.go [hunk 3]
```diff
@@ -106,7 +106,8 @@ func InitNode(cfg *Config) error {
 	r.Stop()
 
 	// TODO: this should be set in the genesis file, not the config
-	if cfg.Core.BabeThreshold == babe.MaxThreshold {
+	// TODO: #1233 confirm
+	if cfg.Core.BabeThresholdNumerator == babe.MaxThreshold {
```
Comment (noot) [inline dot/node.go:110]: you can remove this and directly set the `ThresholdNumerator` and `ThresholdDenominator` in the babe service config

Comment (noot) [reply dot/node.go:110]: which i've seen you've done so you can remove this whole code block

File: dot/services.go [hunk 4]
```diff
@@ -164,17 +164,18 @@ func createBABEService(cfg *Config, rt runtime.LegacyInstance, st *state.Service
...
+		BlockState:           st.Block,
+		StorageState:         st.Storage,
+		TransactionState:     st.Transaction,
+		EpochState:           st.Epoch,
+		StartSlot:            bestSlot + 1,
+		EpochLength:          cfg.Core.EpochLength,
+		ThresholdNumerator:   cfg.Core.BabeThresholdNumerator,
+		ThresholdDenominator: cfg.Core.BabeThresholdDenominator,
+		SlotDuration:         cfg.Core.SlotDuration,
+		Authority:            cfg.Core.BabeAuthority,
```
Comment (noot) [inline dot/services.go:178]: as a side note we should probably rename `core` to `consensus` or something, don't need to do that here though haha

File: lib/babe/babe.go [hunk 5]
```diff
@@ -33,11 +33,16 @@ import (
...
+// TODO: update #1233
 var (
 	// MaxThreshold is the maximum BABE threshold (node authorized to produce a block every slot)
 	MaxThreshold = big.NewInt(0).SetBytes([]byte{0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff})
 	// MinThreshold is the minimum BABE threshold (node never authorized to produce a block)
 	MinThreshold = big.NewInt(0)
+	// DefaultThresholdNumerator is the default threshold numerator
+	DefaultThresholdNumerator = big.NewInt(10)
+	// DefaultThresholdDenominator is the default threshold denominator
+	DefaultThresholdDenominator = big.NewInt(1)
```
Comment (noot) [inline lib/babe/babe.go:45]: you can delete this entire code block including `MaxThreshold` and `MinThreshold`, `MaxThreshold` is equal to 1/1 and `MinThreshold` is equal to 0/1

File: lib/babe/babe.go [hunk 6]
```diff
@@ -185,13 +191,14 @@ func (b *Service) setEpochData(cfg *ServiceConfig, genCfg *types.BabeConfigurati
 		}
 	}
 
-	if cfg.Threshold == nil {
+	if cfg.ThresholdNumerator == nil {
```
Comment (noot) [inline lib/babe/babe.go:194]: check for both numerator and denominator being set here, if so then use numerator as C1 and denominator as C2 as input to `CalculateThreshold`

File: lib/babe/babe.go [hunk 7]
```diff
@@ -185,13 +191,14 @@ func (b *Service) setEpochData(cfg *ServiceConfig, genCfg *types.BabeConfigurati
...
 		b.epochData.threshold, err = CalculateThreshold(genCfg.C1, genCfg.C2, len(b.epochData.authorities))
 		if err != nil {
 			return err
 		}
 	} else {
-		b.epochData.threshold = cfg.Threshold
+		// TODO: #1233 confirm
+		b.epochData.threshold = cfg.ThresholdNumerator.Div(cfg.ThresholdNumerator, cfg.ThresholdDenominator)
```
Comment (noot) [inline lib/babe/babe.go:201]: `CalculateThreshold` is used to calculate the threshold

File: lib/babe/verify_test.go [hunk 8]
```diff
@@ -55,7 +55,8 @@ func TestVerificationManager_OnDisabled_InvalidIndex(t *testing.T) {
 	vm := newTestVerificationManager(t, nil)
 
 	babeService := createTestService(t, &ServiceConfig{
-		Threshold: maxThreshold,
+		ThresholdNumerator:   maxThreshold,
+		ThresholdDenominator: DefaultThresholdDenominator,
```
Comment (noot) [inline lib/babe/verify_test.go:59]: change this to 1 and 1

File: tests/utils/gossamer_utils.go [hunk 9]
```diff
@@ -465,11 +467,12 @@ func generateConfigBabeMaxThreshold() *ctoml.Config {
...
-		Roles:            4,
-		BabeAuthority:    true,
-		GrandpaAuthority: true,
-		BabeThreshold:    "max",
-		SlotDuration:     500,
+		Roles:                    4,
+		BabeAuthority:            true,
+		GrandpaAuthority:         true,
+		BabeThresholdNumerator:   babe.MaxThreshold,
+		BabeThresholdDenominator: babe.DefaultThresholdDenominator,
```
Comment (noot) [inline tests/utils/gossamer_utils.go:474]: change to 1 and 1

File: tests/utils/gossamer_utils.go [hunk 10]
```diff
@@ -506,7 +509,8 @@ func generateConfigNoBabe() *ctoml.Config {
 		SyncLvl:    "debug",
 		NetworkLvl: "debug",
 	}
-	cfg.Core.BabeThreshold = "max"
+	cfg.Core.BabeThresholdNumerator = babe.MaxThreshold
+	cfg.Core.BabeThresholdDenominator = big.NewInt(1)
```
Comment (noot) [inline tests/utils/gossamer_utils.go:513]: change to 0 and 1

File: dot/config.go (hunk 1 above)
Comment (RyRy79261) [reply dot/config.go:90]: I'm not sure, I haven't set up marshaling just yet

File: dot/node.go [hunk 11]
```diff
@@ -105,12 +104,6 @@ func InitNode(cfg *Config) error {
 
 	r.Stop()
 
-	// TODO: this should be set in the genesis file, not the config
-	if cfg.Core.BabeThreshold == babe.MaxThreshold {
-		babeCfg.C1 = 1
-		babeCfg.C2 = 1
-	}
-
```
Comment (noot) [inline dot/node.go:113]: sorry you will need to set C1 and C2 here still, can you check if config numerator and denominator are set, and if so set them to `babeCfg.C1` and `babeCfg.C2` respectively?

File: lib/babe/babe.go [hunk 12]
```diff
@@ -185,13 +179,16 @@ func (b *Service) setEpochData(cfg *ServiceConfig, genCfg *types.BabeConfigurati
 		}
 	}
 
-	if cfg.Threshold == nil {
+	if cfg.ThresholdNumerator == 0 && cfg.ThresholdDenominator == 0 {
```
Comment (noot) [inline lib/babe/babe.go:182]: can you change this to be if `cfg.ThresholdDenominator == 0` only? since we should allow numerator to be 0

**Review by noot — COMMENTED** [review]
looking good!

File: dot/node.go [hunk 13]
```diff
@@ -105,15 +104,15 @@ func InitNode(cfg *Config) error {
...
 	// declare genesis data
 	data := gen.GenesisData()
 
+	// TODO: this should be set in the genesis file, not the config
+	if cfg.Core.BabeThresholdDenominator != 0 && cfg.Core.BabeThresholdNumerator != 0 {
```
Comment (noot) [inline dot/node.go:111]: change to only check `cfg.Core.BabeThresholdDenominator != 0`

File: lib/babe/babe.go [hunk 14]
```diff
@@ -185,13 +179,16 @@ func (b *Service) setEpochData(cfg *ServiceConfig, genCfg *types.BabeConfigurati
...
 		b.epochData.threshold, err = CalculateThreshold(genCfg.C1, genCfg.C2, len(b.epochData.authorities))
 		if err != nil {
 			return err
 		}
 	} else {
-		b.epochData.threshold = cfg.Threshold
+		b.epochData.threshold, err = CalculateThreshold(cfg.ThresholdNumerator, cfg.ThresholdDenominator, len(b.epochData.authorities))
+		if err != nil {
+			return err
+		}
```
Comment (noot) [inline lib/babe/babe.go:191]: nitpick, but you could move the err check below the if/else

**Review by noot — COMMENTED** [review]
looking good!!

File: lib/babe/babe.go (hunk 14 above)
Comment (RyRy79261) [reply lib/babe/babe.go:191]: Ahh yeah should of spotted that, thanks!

**Review by noot — APPROVED** [review]
great work!! tested it locally, everything seems good :D

**Review by arijitAD — APPROVED** [review]
Looks Good!

**Pull request merged**